### ISSUE TITLE
commit 694cdb033214e6baa314ae396fbeff1ce6a77ca0 breaks mnist model on s390x

### ISSUE THREAD

**Issue by gongsu832** [onnx/onnx-mlir#600]
I have a mnist model and driver that worked fine before this patch but now is getting the wrong results. This looks like an endian problem. I remember Tian asked about the endian issue for this patch but there was no response.

Both native library and JNI jar are giving the wrong results (expected since JNI simply calls the native library).

Commands to compile the native library and run the inference:
```
# onnx-mlir --EmitLib model.onnx
# gcc -g -Iinclude_path mnist.c -o mnist model.so -Llib_path -lcruntime
# LD_LIBRARY_PATH=. ./mnist
```
`include_path` and `lib_path` are where runtime headers and libraries are, respectively. You should see the following output:
```
input  signature: [    { "type" : "float" , "dims" : [1 , 1 , 28 , 28]  }

]
output signature:  [   { "type" : "float" , "dims" : [1 , 10]  }

]
prediction[0] = -118029507121385983488744457502720.000000
prediction[1] = -71.833557
prediction[2] = 1230729779023607448758206332928.000000
prediction[3] = 112626513890916027531402412032.000000
prediction[4] = -22.549494
prediction[5] = 9.097518
prediction[6] = 1172105306786765666460368896.000000
prediction[7] = -13.723551
prediction[8] = -2.159079
prediction[9] = -3669.105713
The digit is 2
```
The correct results should be:
```
input  signature: [    { "type" : "float" , "dims" : [1 , 1 , 28 , 28]  }

]
output signature:  [   { "type" : "float" , "dims" : [1 , 10]  }

]
prediction[0] = 84.358864
prediction[1] = -71.825768
prediction[2] = -2.669611
prediction[3] = -23.258825
prediction[4] = -22.675903
prediction[5] = 9.237737
prediction[6] = 18.885262
prediction[7] = -13.772935
prediction[8] = -2.074757
prediction[9] = 0.823274
The digit is 0
```

Commands to compile the JNI jar and run the inference:
```
# onnx-mlir --EmitJNI model.onnx
# javac -cp model.jar MnistTestDriver.java
# ONNX_MLIR_JNI_LOG_LEVEL=debug java -cp .:model.jar MnistTestDriver
```
You should see the following output:
```
[    { "type" : "float" , "dims" : [1 , 1 , 28 , 28]  }

]
 [   { "type" : "float" , "dims" : [1 , 10]  }

]
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:data=[ -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 ... ]
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:shape=[ 1 1 28 28 ]
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:strides=[ 784 784 28 1 ]
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:dataType=1
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:bufferSize=3136
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:rank=4
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:owning=0
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:numElems=784
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:data=[ -118029507121385983488744457502720.000000 -71.833557 1230729779023607448758206332928.000000 112626513890916027531402412032.000000 -22.549494 9.097518 1172105306786765666460368896.000000 -13.723551 -2.159079 -3669.105713 ]
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:shape=[ 1 10 ]
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:strides=[ 10 1 ]
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:dataType=1
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:bufferSize=40
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:rank=2
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:owning=1
[2021-04-10 12:59:13 -0400][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:numElems=10
shape=[1, 10]
prediction[0] = -1.1802951E32
prediction[1] = -71.83356
prediction[2] = 1.2307298E30
prediction[3] = 1.1262651E29
prediction[4] = -22.549494
prediction[5] = 9.097518
prediction[6] = 1.1721053E27
prediction[7] = -13.723551
prediction[8] = -2.159079
prediction[9] = -3669.1057
The digit is 2
```
The `data=[ -118029507121385983488744457502720.000000 ...` line are the results coming back from the native code, which are wrong. The `prediction[0] = -1.1802951E32 ...` lines are printed out by the Java code, which basically verify that it got the same results from the native code.

The correct results should be (before this patch and on amd64):
```
[    { "type" : "float" , "dims" : [1 , 1 , 28 , 28]  }

]
 [   { "type" : "float" , "dims" : [1 , 10]  }

]
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:data=[ -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 -0.424213 ... ]
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:shape=[ 1 1 28 28 ]
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:strides=[ 784 784 28 1 ]
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:dataType=1
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:bufferSize=3136
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:rank=4
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:owning=0
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_java_to_native:246 omt[0]:numElems=784
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:data=[ 84.358864 -71.825768 -2.669611 -23.258825 -22.675903 9.237737 18.885262 -13.772935 -2.074757 0.823274 ]
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:shape=[ 1 10 ]
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:strides=[ 10 1 ]
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:dataType=1
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:bufferSize=40
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:rank=2
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:owning=1
[2021-04-10 17:10:09 +0000][debug]jniwrapper.c:omtl_native_to_java:313 omt[0]:numElems=10
shape=[1, 10]
prediction[0] = 84.358864
prediction[1] = -71.82577
prediction[2] = -2.6696112
prediction[3] = -23.258825
prediction[4] = -22.675903
prediction[5] = 9.237737
prediction[6] = 18.885262
prediction[7] = -13.772935
prediction[8] = -2.074757
prediction[9] = 0.82327354
The digit is 0
```

[mnist.zip](https://github.com/onnx/onnx-mlir/files/6290416/mnist.zip)

**Comment by caoimhinuibrian** [comment]
What is the endianess of the data in the model? I'm guessing little-endian. Is it possible that in the process of writing to a file we converted the constants to big-endian on s390x? @tungld @tjingrant wht do you think?

**Comment by tjingrant** [comment]
With our previous way of handling const pool, they're stored with native endianness. The endianness of the data is recorded in a symbol `constPackIsLE` and checked during parameter loading time to be compatible with native architecture. 

So I think JNI here probably expects native endianness weights. Thus if things go wrong on s390x it'd be my guess to say that weights are stored instead as LE on s390x.

**Comment by gongsu832** [comment]
By convention, all weights on the disk should be in LE and converted to native endian when they are loaded into memory.

**Comment by tungld** [comment]
Quick comment: new way stores weights with native endianness also, similar to the old way (e.g. reading data from raw buffer of dense attributes, and saving it). We have tests for le and be separately here https://github.com/onnx/onnx-mlir/blob/master/test/mlir/krnl/krnl_constant_to_llvm_be/constant.mlir.

**Comment by caoimhinuibrian** [comment]
> By convention (most specs including ONNX), all raw weights on the disk should be in LE and converted to native endian when they are loaded into memory. That was how we handled them in DLC Classic.

They aren't loaded from disk. They are stored in the model in the rodata section.

**Comment by caoimhinuibrian** [comment]
> Quick comment: the new way stores weights with native endianness also, similar to the old way (e.g. reading data from the raw buffer of dense attributes, and saving it). We have tests for le and be separately here https://github.com/onnx/onnx-mlir/blob/master/test/mlir/krnl/krnl_constant_to_llvm_be/constant.mlir.

@tungld could you take a look at @gongsu832's testcase above

**Comment by tungld** [comment]
> > Quick comment: the new way stores weights with native endianness also, similar to the old way (e.g. reading data from the raw buffer of dense attributes, and saving it). We have tests for le and be separately here https://github.com/onnx/onnx-mlir/blob/master/test/mlir/krnl/krnl_constant_to_llvm_be/constant.mlir.
> 
> @tungld could you take a look at @gongsu832's testcase above 


Yes, sure.

**Comment by gongsu832** [comment]
> > By convention (most specs including ONNX), all raw weights on the disk should be in LE and converted to native endian when they are loaded into memory. That was how we handled them in DLC Classic.
> 
> They aren't loaded from disk. They are stored in the model in the rodata section.
> 
> And they should be always stored in native endianness so that normal loads can access them.

Yes I realize that so I was implying there could be problem since now they are stored in the model itself. In a sense, the weights are still stored on disk since the model itself is stored on disk. But I think here we are not dealing with raw weights like we did in DLC Classic, which is just an array of floats stored on disk. rodata should have metadata to tell whether the data is in LE or BE.

With rodata embedded in the .so, it makes sense that it should be in native endian since you can't copy the .so from one arch to another. But with raw weights saved separately, they can be copied around so they must be in LE. I'm guessing the problem here is that the .so on s390x expects data to be stored in BE but the rodata is stored in LE. Here is the comparison of rodata dump on s390x and amd64:

s390x
```
Hex dump of section [13] '.rodata', 24448 bytes at offset 0xcb20:
  0x00000000 ff800000 f4ba37bd 3151ff3b 71788b3d ......7.1Q.;qx.=
  0x00000010 6fb5f53c 877101be 81950f3e 6c7262bd o..<.q.....>lrb.
  0x00000020 af464abd 0ab1ac3d c5655fbd 00000000 .FJ....=.e_.....
...
```
amd64
```
Hex dump of section [14] '.rodata', 24476 bytes at offset 0xc000:
  0x00000000 000080ff f4ba37bd 3151ff3b 71788b3d ......7.1Q.;qx.=
  0x00000010 6fb5f53c 877101be 81950f3e 6c7262bd o..<.q.....>lrb.
  0x00000020 af464abd 0ab1ac3d c5655fbd 00000000 .FJ....=.e_.....
...
```

**Comment by tungld** [comment]
As I remember, when we worked with endianness, if there was a problem with endianness, we often saw a lot of `0` in the result. But in this case it looks different. 

After digging in endianness but could not get a better result, I changed the optimization level of `opt` from `-O3` to `-O2`, and got the correct results:

```bash
tungld@c5868d957a2e:~/dl/onnx-mlir/build$ git diff
diff --git a/src/MainUtils.cpp b/src/MainUtils.cpp
index e7e0015..6a1acb9 100644
--- a/src/MainUtils.cpp
+++ b/src/MainUtils.cpp
@@ -257,7 +257,7 @@ void genLLVMBitcode(const mlir::OwningModuleRef &module,
   // Use the LLVM's 'opt' command to optimize the bitcode.
   string optPath = getToolPath("opt");
   Command optBitcode(/*exePath=*/!optPath.empty() ? optPath : kOptPath);
-  optBitcode.appendStr("-O3")
+  optBitcode.appendStr("-O2")
       .appendStr(getTargetOptions())
       .appendList({"-o", optimizedBitcodePath})
       .appendStr(unoptimizedBitcodePath)
```

```tungld@c5868d957a2e:~/dl/dlc-lab/mnist$ ./mnist
input  signature: [    { "type" : "float" , "dims" : [1 , 1 , 28 , 28]  }

]
output signature:  [   { "type" : "float" , "dims" : [1 , 10]  }

]
prediction[0] = 84.358864
prediction[1] = -71.825768
prediction[2] = -2.669611
prediction[3] = -23.258825
prediction[4] = -22.675903
prediction[5] = 9.237737
prediction[6] = 18.885262
prediction[7] = -13.772935
prediction[8] = -2.074757
prediction[9] = 0.823274
The digit is 0
```

@gongsu832 could you please try to use `-O2` for `opt` and see if the issue happens again? Thanks`

**Comment by gongsu832** [comment]
@tungld yes changing to -O2 gives the correct results. The .rodata dump on s390x now looks like this:
```
Hex dump of section [13] '.rodata', 24464 bytes at offset 0xbe50:
  0x00000000 ff800000 00000000 00000000 00000000 ................
  0x00000010 bd37baf4 3bff5131 3d8b7871 3cf5b56f .7..;.Q1=.xq<..o
  0x00000020 be017187 3e0f9581 bd62726c bd4a46af ..q.>....brl.JF.
...
```
But it's just too weird to have a different optimization level writing data in different endian. Do we know what exactly -O2 and -O3 does for opt? It sounds like opt is broken.

**Comment by tungld** [comment]
Yes. It appears that opt is broken. I don't know what optimization in opt caused the issue. I did a check on inference performance comparision and saw no different between -O3 and -O2 for resnet50 at this moment. Perhaps, we should use -O2 for safety.

**Comment by gongsu832** [comment]
Looks like we have no choice but to use -O2 until opt is fixed. I guess we should report this to LLVM?

**Comment by tungld** [comment]
> I guess we should report this to LLVM?

Yes, I think so.

**Comment by tjingrant** [comment]
Just to make sure I understand the issue, so opt flips the byte order of global string in LLVM when -O3 flag is enabled?

That means if we write a helloworld program like this:

```cpp
#include <iostream>

const string globalStr = "Hello World";

int main() {
    std::cout << globalStr;
    return 0;
}
```

And compile with -O3 on s390x, it should print out garbage?

**Comment by tjingrant** [comment]
@gongsu832 do you know why this issue slipped through our CI? Is it because this is something zOS specific, and therefore is not detectable under zLinux?

**Comment by gongsu832** [comment]
@tjingrant Endian is not a problem with strings since they are interpreted byte-by-byte. Endian is only a problem when you interpret data type with more than one byte. For example, if you have an array of shorts stored in memory like this:
```
0x01 0x02 0x03 0x04 ...
```
If you print out the shorts on BE, you will get `0x0102`, `0x0304`. However, if you print out the shorts on LE, you will get `0x0201`, `0x0403`. This is a property of the architecture, not related to OS. So it doesn't matter whether it's z/OS or zLinux.

I'm also wondering why the problem didn't happen with the models in "make test" like vgg, etc. Unless the models in the test don't have weight constants. But surely that's not true?

**Comment by caoimhinuibrian** [comment]
do we have any actual evidence that this is an endianness problem? Perhaps it's just an O3 bug exposed by the change to put the constants in the LLVM IR?

**Comment by gongsu832** [comment]
@caoimhinuibrian That was what we suspected, that the -O3 has some bug that caused the .rodata on s390x to be written out in LE order (except the first word). Comparing to -O2, -O3 also seemed to trim off a few zero bytes (24464 vs 24448). Maybe it's doing some packing and forgets to check the endian and defaults to LE.

**Comment by gongsu832** [comment]
OK after comparing a few models with -O2 and -O3, it looks like the -O3 "bug" is definitely related to the size of the rodata. It only happens with small rodata, which explains why our CI didn't catch it because all the models used in the CI have large rodata.

From the models I have, -O3 rodata is bad when its size is below 31KB, but good when its size is above 72KB. So I venture to say that the threshold is either 32KB or 64KB, below which the "bug" will manifest. So I think what Tian said during the meeting that smaller rodata go through a different treatment (not through protobuf which would have proper endian handling) makes a lot of sense.

**Comment by tungld** [comment]
@gongsu832 good finding! Thanks!

I generated LLVMIR from the mnist model, using `onnx-mlir -EmitLLVMIR`, then translated to LLVM using `mlir-translate -mlir-to-llvmir` to have an output, say, `mnist.ll`.

Now we play with bicode optimization. The idea is 1) generating bitcode from LLVM 2) optimizing bitcode using -O2 and -O3, then 3) converting back to LLVM.

Comparing the LLVM outputs of -O2 and -O3, I see that:
- One constant was disappeared, that was `@constant_7 = internal unnamed_addr constant [40 x i8] c"\BD7\BA\F4;\FFQ1=\8Bxq<\F5\B5o\BE\01q\87>\0F\95\81\BDbrl\BDJF\AF=\AC\B1\0A\BD_e\C5"`
- Other constants are the same and similar to those in the input LLVM.

The missing of the constant looks consistent with the wrong inference output where some result values are correct and some result values are very very large (undefined values), as Gong showed:
```
prediction[0] = -118029507121385983488744457502720.000000
prediction[1] = -71.833557
prediction[2] = 1230729779023607448758206332928.000000
prediction[3] = 112626513890916027531402412032.000000
prediction[4] = -22.549494
prediction[5] = 9.097518
prediction[6] = 1172105306786765666460368896.000000
prediction[7] = -13.723551
prediction[8] = -2.159079
prediction[9] = -3669.105713
The digit is 2
```

So, it seems the problem is not endian but somehow -O3 removed our constant.

I attached the LLVM input and outputs here.
[mnist-llvm.zip](https://github.com/onnx/onnx-mlir/files/6302396/mnist-llvm.zip)

**Comment by tungld** [comment]
One more observation about the mnist model. There were two dead constants generated by lowering onnx.reshape. `-O2` correctly eliminated the two constants, but `-O3` eliminated one more constant and caused the issue.

**Comment by gongsu832** [comment]
@tungld Thanks for the investigation. I think the problem is both, endian and missing constants. If you compare the rodata dump between -O3 and -O2:

-O3
```
Hex dump of section [13] '.rodata', 24448 bytes at offset 0xcb20:
  0x00000000 ff800000 f4ba37bd 3151ff3b 71788b3d ......7.1Q.;qx.=
  0x00000010 6fb5f53c 877101be 81950f3e 6c7262bd o..<.q.....>lrb.
  0x00000020 af464abd 0ab1ac3d c5655fbd 00000000 .FJ....=.e_.....
...
```
-O2
```
Hex dump of section [13] '.rodata', 24464 bytes at offset 0xbf70:
  0x00000000 ff800000 00000000 00000000 00000000 ................
  0x00000010 bd37baf4 3bff5131 3d8b7871 3cf5b56f .7..;.Q1=.xq<..o
  0x00000020 be017187 3e0f9581 bd62726c bd4a46af ..q.>....brl.JF.
...
```
You can clearly see that -O3 is missing all the 0s after `ff800000` and the rest of the constants are in LE, e.g., -O3 stores a `f4ba37bd` while -O2 stores `bd37baf4`, etc.

**Comment by tungld** [comment]
It seems to me endianness in the generated -O3 bitcode is correct since we can obtain correct constants when they are converted back to LLVM. Any possibility that the endianness problem happened on the way from bitcode to .so?

**Comment by gongsu832** [comment]
If the endianness is still correct in the bitcode, then I guess it must have happened on the way from bitcode to .so. Unfortunately I don't know much about LLVM and what's involved going from bitcode to .so (beyond the fact that llc was used to do the compilation). Hopefully others @tjingrant @caoimhinuibrian etc. can chime in and offer their insights.

**Comment by seroyer** [comment]
I believe that the issue no longer shows up on the master branch.  However, it still shows up on the llvm-12-release-branch which also has the change to use -O2.  So I don't believe that -O2 is sufficient.

**Comment by caoimhinuibrian** [comment]
@seroyer I do not observe this problem on the llvm-12-release-branch

**Comment by seroyer** [comment]
I still observe the problem on s390x.  Maybe it is an endian issue?

**Comment by seroyer** [comment]
My test client running the model zoo tests worked before the changes to the embedded consts.  It works correctly on master branch.  But it fails on release branch:

```
$ /code/client/bin/modelzoo -f mnist/mnist.tests --validate
Output does not match expected within tolerance.  Element index 0, values -118029507121385983488744457502720.000000 and 975.670105.
Output does not match expected within tolerance.  Element index 2, values 1230729779023607448758206332928.000000 and 6574.568359.
Output does not match expected within tolerance.  Element index 3, values 112626513890916027531402412032.000000 and 668.028931.
Output does not match expected within tolerance.  Element index 4, values -917.145142 and -917.270935.
Output does not match expected within tolerance.  Element index 5, values -1671.776123 and -1671.635864.
Output does not match expected within tolerance.  Element index 6, values 1172105306786765666460368896.000000 and -1952.759888.
Output does not match expected within tolerance.  Element index 7, values -61.499847 and -61.549873.
Output does not match expected within tolerance.  Element index 8, values -777.260559 and -777.176636.
Output does not match expected within tolerance.  Element index 9, values -5109.459961 and -1439.531616.
Output validation failed on test number 0
```

Again, this is running on s390x, so big endian.

**Comment by tungld** [comment]
Hi,

Just found that `loop-unroll` in `opt` was the root cause. It unrolled a loop that added two arrays of 10 elements. One of the array was constant. During the unroll, it seemed it used little endian or wrongly read the array.
```
%652 = load float, float* %93, align 4, !dbg !253
  %653 = fadd float %652, 0xC69746F7A0000000, !dbg !254
  store float %653, float* %93, align 4, !dbg !255
  %654 = getelementptr i8, i8* %11, i64 4, !dbg !256
  %655 = bitcast i8* %654 to float*, !dbg !256
  %656 = load float, float* %655, align 4, !dbg !253
  %657 = fadd float %656, 0x3E2A3FE760000000, !dbg !254
  store float %657, float* %655, align 4, !dbg !255
  %658 = getelementptr i8, i8* %11, i64 8, !dbg !256
  %659 = bitcast i8* %658 to float*, !dbg !256
  %660 = load float, float* %659, align 4, !dbg !253
  %661 = fadd float %660, 0x462F1167A0000000, !dbg !254
  store float %661, float* %659, align 4, !dbg !255
  %662 = getelementptr i8, i8* %11, i64 12, !dbg !256
  %663 = bitcast i8* %662 to float*, !dbg !256
  %664 = load float, float* %663, align 4, !dbg !253
  %665 = fadd float %664, 0x45F6BEA780000000, !dbg !254
  store float %665, float* %663, align 4, !dbg !255
  ```

I enabled -O3, and just turned `loop-unroll` off:
```
tungld@c5868d957a2e:~/dl/onnx-mlir/build$ git diff
diff --git a/src/MainUtils.cpp b/src/MainUtils.cpp
index 7791b22..9bfa53d 100644
--- a/src/MainUtils.cpp
+++ b/src/MainUtils.cpp
@@ -261,8 +261,9 @@ void genLLVMBitcode(const mlir::OwningModuleRef &module,
   // Use the LLVM's 'opt' command to optimize the bitcode.
   string optPath = getToolPath("opt");
   Command optBitcode(/*exePath=*/!optPath.empty() ? optPath : kOptPath);
-  optBitcode.appendStr("-O2")
+  optBitcode.appendStr("-O3")
       .appendStr(getTargetOptions())
+      .appendStr("--disable-loop-unrolling")
       .appendList({"-o", optimizedBitcodePath})
       .appendStr(unoptimizedBitcodePath)
       .exec();
```
and mnist worked well:
```
prediction[0] = 84.358871
prediction[1] = -71.825768
prediction[2] = -2.669612
prediction[3] = -23.258831
prediction[4] = -22.675907
prediction[5] = 9.237735
prediction[6] = 18.885262
prediction[7] = -13.772935
prediction[8] = -2.074757
prediction[9] = 0.823274
The digit is 0
```

**Comment by caoimhinuibrian** [comment]
Great! is the problem the constant here?
%653 = fadd float %652, 0xC69746F7A0000000, !dbg !254
it does look backwards ....
How should we proceed with this? Can we turn of unrolling for now and report the bug to llvm?

**Comment by caoimhinuibrian** [comment]
... or should we revert the earlier constant fix (putting constants in the IR) and go back to binding the constants separately until we can get llvm to look at this? @gongsu832 @tjingrant @AlexandreEichenberger , any thoughts?

**Comment by tungld** [comment]
It seems that loop unrolling is only applied to small loops (constant and small trip count): https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/IPO/PassManagerBuilder.cpp#L466. This looks consistent with what Gong observed.

**Comment by gongsu832** [comment]
Yeah @tungld great detective work! I think it would be nice to keep this patch since it fixes the problem of having to copy constants buffer. So I would vote for keeping the patch and turning off loop unroll for now, and pointing the gun at the LLVM folks. We do also need to remember that this patch increases the memory footprint (and causes CircleCI build to fail) but I think Tung san is going to look into that.

**Comment by AlexandreEichenberger** [comment]
Good find @tungld

**Comment by tjingrant** [comment]
Impressive findings @tungld ! 

@caoimhinuibrian I think we need to evaluate the impact of turning loop unrolling off, probably by running a few representative benchmarks. I wonder if @AlexandreEichenberger knows the effect of switching on/off loop unrolling in LLVM?

It does appear that hot loops should have been unrolled already at the level of LLVM, so maybe nothing to be weary about.

**Comment by caoimhinuibrian** [comment]
Looked some more into this:
The constant is correctly stored as big-endian on Z. The relevant constant is constant_7
@constant_7 = internal unnamed_addr constant [40 x i8] c"\BD7\BA\F4;\FFQ1=\8Bxq<\F5\B5o\BE\01q\87>\0F\95\81\BDbrl\BDJF\AF=\AC\B1\0A\BD_e\C5"

On x86 this is@constant_7 = internal unnamed_addr constant [40 x i8] c"\F4\BA7\BD1Q\FF;qx\8B=o\B5\F5<\87q\01\BE\81\95\0F>lrb\BD\AFFJ\BD\0A\B1\AC=\C5e_\BD"

as you can see, these are in opposite endiannesses.

The bug first appears in the *** IR Dump After Recognize loop idioms *** what appears to happen is that the optimization doesn't realize that the constant is in big-endian format and puts a literal that should represent the float into the instruction
  %587 = getelementptr float, float* %13, i64 %585
  %588 = load float, float* %587, align 4
  %589 = fadd float %588, 0xBFA6F75E80000000
  store float %589, float* %587, align 4
As Tung discovered, the loop has to be unrolled to allow this to be done. It took a while to figure out what these constants are (it does the same thing on x86, but there the values are correct). The constant is the hex value in 64 bit float that represents the actual 32 bit quantity, that's why they don't look the same in hex.

I think it's possible that this optimization is failing because Tung's fix to the constant issue pretends that the float array is actual an array of char, and the optimizer gets confused. I'm not sure what to do about it. What was the original problem with passing constants to llvm that caused us to pull them out into a file? I think it was compile-time performance?

**Comment by tungld** [comment]
> What was the original problem with passing constants to llvm that caused us to pull them out into a file? I think it was compile-time performance?

Yes, it was the compile-time performance, coming from using DenseElementsAttr.

I also think the optimizer gets confused with our float array. One simple way for this problem is using a threshold to choose between using a float array (StringAttr) or DenseElementsAttr. We only use the float array if the number of elements is greater than the threshold. With a small threshold, it would not affect compile-time performance much. The important thing is how small the threshold is? In Resnet50, there is a similar loop as mnist for bias add of 1000 elements, and Resnet50 has no issue. Maybe 1024 for the threshold looks fine.

**Comment by AlexandreEichenberger** [comment]
I added the `--disable-loop-unrolling ` option but it does not like both `-O3` and `--disable-loop-unrolling`. Not sure how you got it to work. In any case, this does not change the (unrelated) breakage of my inception test (unrelated)

```
opt --disable-loop-unrolling -O3 -o ./inception_v1.bc ./inception_v1.unoptimized.bc
opt: Unknown command line argument '--disable-loop-unrolling -O3'.  Try: 'opt --help'
opt: Did you mean '--disable-loop-unrolling'?
opt --disable-loop-unrolling -O3 -o ./inception_v1.bc ./inception_v1.unoptimized.bc
```

**Comment by caoimhinuibrian** [comment]
fixed by #639

**Comment by tungld** [comment]
Just a memo. This PR #819 annotates ModuleOp with endianness information. With this information, LLVM optimizations correctly interprets our float array which is stored in string. So I can say we completely solve this issue on Z.

**Comment by gongsu832** [comment]
@tungld Kudos for getting to the root of this.
